src/data/transcript.rs

```diff
@@ -0,0 +1,51 @@
+use core::result::Result;
+use std::path::PathBuf;
+
+use crate::SharedTranscript;
+use serde::{de::DeserializeOwned, ser::Serialize};
+
+pub trait Contribution: Serialize + DeserializeOwned {
+    type Receipt: Serialize;
+    fn get_receipt(&self) -> Self::Receipt;
+}
+
+pub trait Transcript: Serialize + DeserializeOwned {
+    type ContributionType: Contribution;
+    type ValidationError: Serialize;
+
+    fn verify_contribution(
+        &self,
+        contribution: &Self::ContributionType,
+    ) -> Result<(), Self::ValidationError>;
+
+    fn update(&self, contribution: &Self::ContributionType) -> Self;
+
+    fn get_contribution(&self) -> Self::ContributionType;
+}
+
+pub async fn read_transcript_file<T: DeserializeOwned + Send + 'static>(path: PathBuf) -> T {
+    let handle = tokio::task::spawn_blocking::<_, T>(|| {
```

Comment:
I usually just AsynRead/Write into a buffer and then serde_json from the buffer. This works too. Not sure what pros/cons are of either approach.

Reply:
My hope was to do the deserialization on the fly, without having to materialize the whole buffer, looks like serde can do this